cmd/agent/main.go

```diff
@@ -205,29 +206,25 @@ func transportSettings(apiEndpoint string) (string, bool) {
 	return target, insecure
 }
 
-func startHeartbeat(taskId int64, clientToken string) {
-	sendHeartbeat(taskId, clientToken)
-	go heartbeat(taskId, clientToken)
-}
-
-func heartbeat(taskId int64, clientToken string) {
-	ticker := time.NewTicker(60 * time.Second)
-	for {
-		sendHeartbeat(taskId, clientToken)
-		<-ticker.C
-	}
-}
-
-func sendHeartbeat(taskId int64, clientToken string) {
+func runHeartbeat(taskId int64, clientToken string, conn *grpc.ClientConn) {
 	taskIdentification := api.TaskIdentification{
 		TaskId: taskId,
 		Secret: clientToken,
 	}
-	log.Println("Sending heartbeat...")
-	_, err := client.CirrusClient.Heartbeat(context.Background(), &api.HeartbeatRequest{TaskIdentification: &taskIdentification})
-	if err != nil {
-		log.Printf("Failed to send heartbeat: %v", err)
-	} else {
-		log.Printf("Sent heartbeat!")
+	ticker := time.NewTicker(60 * time.Second)
+	for {
+		log.Println("Sending heartbeat...")
+		_, err := client.CirrusClient.Heartbeat(context.Background(), &api.HeartbeatRequest{TaskIdentification: &taskIdentification})
+		if err != nil {
+			log.Printf("Failed to send heartbeat: %v", err)
+			connectionState := conn.GetState()
+			log.Printf("Connection state: %v", connectionState.String())
+			if connectionState == connectivity.TransientFailure {
+				conn.ResetConnectBackoff()
+			}
+		} else {
+			log.Printf("Sent heartbeat!")
+		}
+		<-ticker.C
```

Comment:
You can just use `time.Sleep()` here since there's no `select`.

Reply:
Fixed in 3efbd18